Implements --dump-shellcompletions

dan-da opened the pull request:
Addresses #73 for auto-completion of commands in a shell.

Supports shells:
  bash, fish, zsh, powershell, elvish

The completions are automatically generated from the values in
structopt, so they stay current as the cli params change over time.

To use the completions, eg in bash:

    safe --dump-completions=bash > /tmp/out && source /tmp/out

**Considerations**

1. --dump-completions works for me, but could be renamed to anything, eg --completions is shorter.

2. I didn't use a shortopt (yet).   Can add if wanted.

3. All the logic is contained in a single if statement.. I could move it into a function if desired.

4. Nothing is specific to safe-cli app.  If there is a common code shared between the apps under safe-api, then I could make/move a func into the common area to be called by each of them, so each would have --dump-completions.  Each would need a modification to structopt, and a single fn call.

5. Does *not* presently support completion of dynamic text such as a URL that user may be entering.  See discussion in #73.

6. To be most useful, cargo install (or whatever installer) should have a step to dump completions and install for user.   Exactly how best to do that with would need a bit of research, especially when dealing with various shells and OSes.  For now, this PR lays necessary groundwork.

<!--
#### Thank you for contributing!

Please reference the issue(s) which this pull request addresses using [keywords](https://help.github.com/articles/closing-issues-using-keywords/) such as:

```
Resolves #452
Fixes #363
Closes #408
```

---

Provide QA team and reviewer steps to test the resolution.
For example:

```
QA:
Easiest way to test this PR would be to:
- Start authd
- login
- safe wallet insert <problem link>
- Expect to see an authentication error output

```

---

Commit messages should conform to the format:

```
<type>(<scope>): <description>

[optional body]

```

For example:

```
fix(auth): proper values returned on auth_decode_ipc_msg errors

  - Test case for authenticator error -208 IncompatibleMockStatus
  - Test case for authenticator error -103 when decoding share MData
    request for non-existent MData

```

Commit `type` can be one of:
**feat**: New feature
**fix**: Bug fix
**docs**: Documentation only changes
**style**: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
**refactor**: A code change that neither fixes a bug nor adds a feature
**perf**: A code change that improves performance
**test**: Adding missing tests
**chore**: Changes to the build process or auxiliary tools and libraries such as documentation generation
**revert**: Reverting a feature, fix, or commit which introduces a regression or new bug

Commit `scope` is open to any succinct term which indicates the effected feature or component.

See [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)

---
Write your description below this line: -->

bochaco commented on safe-cli/cli.rs:
```diff
@@ -68,6 +71,23 @@ pub fn run_with(cmd_args: &[&str], mut safe: &mut Safe) -> Result<(), String> {
+        // Note: here we use name of currently running executable.
+        // Possibly could use value of [[bin]].name in Cargo.toml instead
+        let exec_name = std::env::current_exe()
+            .expect("Can't get the exec path")
```
We need to handle the error is as it will panic otherwise, this can be achieved with `std::env::current_exe().map_err(|err| format!("Can't get the exec path: {}", err))?`

dan-da replied:
now handles all errors.

bochaco commented on safe-cli/cli.rs:
```diff
@@ -68,6 +71,23 @@ pub fn run_with(cmd_args: &[&str], mut safe: &mut Safe) -> Result<(), String> {
+        let exec_name = std::env::current_exe()
+            .expect("Can't get the exec path")
+            .file_name()
+            .expect("Can't get the exec name")
```
The error here also needs to handled so it cannot panic.

bochaco commented on safe-cli/cli.rs:
```diff
@@ -68,6 +71,23 @@ pub fn run_with(cmd_args: &[&str], mut safe: &mut Safe) -> Result<(), String> {
+        // Generates shell completions for <shell> and prints to stdout
+        CmdArgs::clap().gen_completions_to(exec_name, shell, &mut std::io::stdout());
+
+        return Ok(());
```
I think we don't need to return here, it's a flag so it should coexist with other commands and flags. If we don't want it to coexist with other flags then we need it to be a subcommand, or a flag of a subcommand. This is related to a previous comment I made about perhaps thinking of some subcommand which can encapsulate these type of maintenance or setup operations, e.g. `$ safe setup --dump-completions <shell>`.

dan-da replied:
I don't think it makes sense to have it as a flag without returning and then continue on to possibly process other command(s).  eg if I ran safe --dump-completions bash cat <url>, it would prepend before the cat output.

I will see about implementing a setup command.

dan-da replied:
working on setup cmd, but I will just note a counter argument, which is that --help works like --dump-completions in that it returns immediately.  eg *safe --help shell* never enters the shell, so there is precedent.

btw, I am trying to implement it like:

```
safe setup completions dump bash --> bash dump
safe setup completions list --> list of available shells completions.
```

bochaco replied:
The exception proves the rule ;)  seriously, I think --help/-h is definitely the exception and no other case shall exist, I believe, otherwise you may end up in situations where you don't know which flag was applied and which wasn't since one of them was applied and exited.

I think we can handle it with just `$ safe setup completions bash` and if we do something similar to what we did with `--xorurl` flag we can just have `$ safe setup completions --help` to give you what's supported in the help menu. 

We don't have commands or flags for documentation, that's what the help is for, in fact there is a way to have --help to return additional information than -h: https://docs.rs/structopt/0.2.14/structopt/#help-messages, and we were thinking of adding much more help there, even some examples and steps like the user guide but when you do --help.

We should have done the same for the `--output` now that I think of.

dan-da replied:
implemented as `safe setup completions <shell>` per request.

dan-da commented:
Ok, I've rebased and re-worked this quite a lot.  Well, before your last comment I already had safe setup completions <cmd> partially working, so I decided to finish that up first and I added a dumpall cmd.  If nothing else, at least I think this work provides a nice infrastructure in the setup.rs file so that setup commands can easily be 2 levels deep if needed, and for adding other commands.

Regarding the changes you suggested, I'm at a good point now where I could refactor to do it that way, though will likely have to wait until after thxgiving weekend.  Anyway, I wanted to run this implementation by you first.  If you hate the completion <cmd> approach, that's ok.  might be overkill.

One thing I will point out regarding '--json setup completions list' is that the output is machine readable eg by an installer, whereas I think output of --help is not so much.  maybe not a concern, especially if dumpall equivalent remains.

Here's some example usage:

```
$ safe help setup
SUBCOMMANDS:
    completions    Work with shell completions for: [bash, fish, zsh, powershell, elvish]
```

```
$ safe help setup completions
SUBCOMMANDS:
    dump       Dump shell completions.
    dumpall    Dump completions for all shells.
    help       Prints this message or the help of the given subcommand(s)
    list       List shells that have completions available.
```

Each completions sub-command can output either pretty or serialized.  eg:

```
$ safe setup completions list
+------------+
| bash       |
+------------+
| elvish     |
+------------+
| fish       |
+------------+
| powershell |
+------------+
| zsh        |
+------------+
```

```
$ safe setup completions list --json
[
  "bash",
  "elvish",
  "fish",
  "powershell",
  "zsh"
]
```

bochaco commented:
We don't need to rush it @dan-da , so no worries.

> whereas I think output of --help is not so much. maybe not a concern, especially if dumpall equivalent remains.

I think an install script would go the other way around, it would try to detect which shell is supported in the target OS, and then dump its completion passing the shell as arg, so still not sure we need this type of information about CLI arguments to be returned as part of any command really.

As per dump and dumpall, i would make dumpall to be the default when no shell name is provided as arg to the dump command, e.g.:
`$ safe setup completions dump` => dump all
`$ safe setup completions dump bash` => dump bash only

We could even remove the dump argument, unless we already envision more subcommands of the `completions` subcommand, e.g.:
`$ safe setup completions` => dump all
`$ safe setup completions bash` => dump bash only
I'd vote for this to keep the level of subcommands as shallow as possible.

dan-da commented:
On Linux at least, a user can have multiple shells installed, though probably not so common and most will just have bash.

In thinking about all this and the big picture, I go back to my original goal, which was just to add a simple, consistent way to dump completions for each of the safe apps, eg: safe-cli, safe-vault, safe-authd, etc.

Looking at safe_vault usage, I notice it does not presently accept any commands, only flags/options.  safe-authd does accept commands. 

Implementing this as a setup sub-command has been a good exercise for me as I learn rust, but I'm not convinced it's the best approach with the above goal in mind.   Would we add 'setup completions' to safe_vault, safe-authd?  Seems kind of heavy for such a simple thing.

So I go back to thinking we should make it either *--completions <shell>* option or *completions <shell>* command and have it work the same in all CLI apps (and maybe gui apps eventually).

There are not very many available shell options and they can be displayed in the help text for the option/command.

what do you think?

bochaco commented:
yes, if you want to have consistency then perhaps as you say I don't see many more option around this command so simply `$ safe completions` (dump all) and `$ safe completions bash` could do it just fine

dan-da commented:
ok, I implemented `$ safe completions` (dump all) and `$ safe completions bash`.

If this version is acceptable, I could also see about implementing it for safe_vault and safe_authd.  any issues with that?

```
$ safe setup --help
...
SUBCOMMANDS:
    completions    Dump shell completions.
    help           Prints this message or the help of the given subcommand(s)
```

```
$ safe setup completions --help
safe-setup-completions 0.6.0
bochaco <gabrielviganotti@gmail.com>, Josh Wilson <joshuef@gmail.com>, Calum Craig <calum.craig@maidsafe.net>, Chris
O'Neil <chris.oneil@gmail.com>
Dump shell completions.

USAGE:
    safe setup completions [FLAGS] [OPTIONS] [shell]

FLAGS:
    -n, --dry-run    Dry run of command. No data will be written. No coins spent
    -h, --help       Prints help information
        --json       Sets JSON as output serialisation format (alias of '--output json')
    -V, --version    Prints version information

OPTIONS:
        --endpoint <endpoint>     Endpoint of the Authenticator daemon where to send requests to. If not provided,
                                  https://localhost:33000 is assumed.
    -o, --output <output_fmt>     Output data serialisation: [json, jsoncompact, yaml]
        --xorurl <xorurl_base>    Base encoding to be used for XOR-URLs generated. Currently supported: base32z
                                  (default), base32 and base64

ARGS:
    <shell>    one of: [bash, fish, zsh, powershell, elvish]  default = all shells
```

bochaco commented:
I think it's good for a start, I think we can merge this and see how things evolve in terms of other subcommands of `setup` are required or not, if there are no more subcommands that we can add to the `setup` subcommand we can then flatten it up a bit and move `completions` to be a direct subcommand of safe.
